libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m

```diff
@@ -853,17 +853,59 @@ - (NSString *)scopeQueryParamString {
  */
-
 - (void) updateCredentials:(NSDictionary *) params {
+    NSMutableDictionary *changeSet = [NSMutableDictionary new];
+    if (params[kSFOAuthAccessToken]) {
+        if(![params[kSFOAuthAccessToken] isEqualToString:self.credentials.accessToken])
+            changeSet[kSFOAuthAccessToken] = @[self.credentials.accessToken?
+                    self.credentials.accessToken:[NSNull null],
+                    params[kSFOAuthAccessToken]];
+
+        self.credentials.accessToken = params[kSFOAuthAccessToken];
+    }
+
+    if (params[kSFOAuthIssuedAt]) {
+        self.credentials.issuedAt = [[self class] timestampStringToDate:params[kSFOAuthIssuedAt]];
+    }
+
+    if (params[kSFOAuthInstanceUrl]) {
+        NSURL *newInstanceUrl = [NSURL URLWithString:params[kSFOAuthInstanceUrl]];
+
+        if(![newInstanceUrl isEqual:self.credentials.instanceUrl])
+            changeSet[kSFOAuthInstanceUrl] = @[self.credentials.instanceUrl?
+                    [self.credentials.instanceUrl absoluteString]:[NSNull null],
+                    params[kSFOAuthInstanceUrl]];
+
+        self.credentials.instanceUrl = newInstanceUrl;
+    }
 
-   
-    if (params[kSFOAuthAccessToken]) self.credentials.accessToken = params[kSFOAuthAccessToken];
-    if (params[kSFOAuthIssuedAt]) self.credentials.issuedAt = [[self class] timestampStringToDate:params[kSFOAuthIssuedAt]];
-    if (params[kSFOAuthInstanceUrl]) self.credentials.instanceUrl = [NSURL URLWithString:params[kSFOAuthInstanceUrl]];
-    if (params[kSFOAuthId]) self.credentials.identityUrl = [NSURL URLWithString:params[kSFOAuthId]];
-    if (params[kSFOAuthCommunityId]) self.credentials.communityId = params[kSFOAuthCommunityId];
-    if (params[kSFOAuthCommunityUrl]) self.credentials.communityUrl = [NSURL URLWithString:params[kSFOAuthCommunityUrl]];
+    if (params[kSFOAuthId]) {
+        NSURL *newIdUrl = [NSURL URLWithString:params[kSFOAuthId]];
+        if(![newIdUrl isEqual:self.credentials.identityUrl])
+            changeSet[kSFOAuthId] = @[self.credentials.identityUrl?
+                    [self.credentials.identityUrl absoluteString]:[NSNull null],
+                    params[kSFOAuthId]];
 
+        self.credentials.identityUrl = newIdUrl;
+    }
+
+    if (params[kSFOAuthCommunityId]) {
+        if(![changeSet[kSFOAuthCommunityId] isEqualToString:self.credentials.communityId])
+            changeSet[kSFOAuthCommunityId] = @[self.credentials.communityId?
+                    self.credentials.communityId:[NSNull null],
+                    params[kSFOAuthCommunityId]];
+        self.credentials.communityId = params[kSFOAuthCommunityId];
+    }
+
+    if (params[kSFOAuthCommunityUrl]) {
+        NSURL *newCommUrl = [NSURL URLWithString:params[kSFOAuthCommunityUrl]];
+        if(![newCommUrl isEqual:self.credentials.communityUrl])
+            changeSet[kSFOAuthCommunityUrl] = @[self.credentials.communityUrl?
+                    [self.credentials.communityUrl absoluteString]:[NSNull null],
```

Comment:
I'm wondering why you need to pass `NSNull` at all?  Instead of using a dictionary with string constant keys, and a nested array whose position represents old/new values (if I'm reading this code correctly), it might be better to write this up as a proper model object.

```
@interface SFOAuthCredentialChange : NSObject

@property (nonatomic, copy, readonly) NSString *type;
@property (nonatomic, strong, readonly) id oldValue;
@property (nonatomic, strong, readonly) id newValue;

@end
```

Then it could be used as: `NSArray<SFOAuthCredentialChange*> *changes`.  Or something along those lines?  It would certainly make it easier because then you wouldn't have to compare against `[NSNull null]`, you can override `isEqual` as needed, etc.